geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/UserFunctionExecution.java

```diff
@@ -31,138 +34,149 @@
-        List<String> resultMessage = new ArrayList<>();
-        boolean functionSuccess = true;
-
-        if (results != null) {
-          for (Object resultObj : results) {
-            if (resultObj != null) {
-              if (resultObj instanceof Exception) {
-                resultMessage.add(((Exception) resultObj).getMessage());
-                functionSuccess = false;
-              } else {
-                resultMessage.add(resultObj.toString());
-              }
+      if (functionArgs != null && functionArgs.length > 0) {
+        execution = execution.setArguments(functionArgs);
+      }
+      if (filters.size() > 0) {
+        execution = execution.withFilter(filters);
+      }
+
+      List<Object> results = (List<Object>) execution.execute(function.getId()).getResult();
+      List<String> resultMessage = new ArrayList<>();
+      boolean functionSuccess = true;
+
+      if (results != null) {
+        for (Object resultObj : results) {
+          if (resultObj != null) {
+            if (resultObj instanceof Exception) {
+              resultMessage.add(((Exception) resultObj).getMessage());
+              functionSuccess = false;
+            } else {
+              resultMessage.add(resultObj.toString());
             }
           }
         }
-        context.getResultSender().lastResult(
-            new CliFunctionResult(member.getId(), functionSuccess, resultMessage.toString()));
-
-      } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
-        context.getResultSender()
-            .lastResult(new CliFunctionResult(member.getId(), false,
-                CliStrings.format(
-                    CliStrings.EXECUTE_FUNCTION__MSG__RESULT_COLLECTOR_0_NOT_FOUND_ERROR_1,
-                    resultCollectorName, e.getMessage())));
-      } catch (Exception e) {
-        context.getResultSender().lastResult(
-            new CliFunctionResult(member.getId(), false, "Exception: " + e.getMessage()));
-      } finally {
-        securityService.logout();
       }
+      context.getResultSender().lastResult(
+          new CliFunctionResult(member.getId(), functionSuccess, resultMessage.toString()));
 
-    } catch (Exception ex) {
+    } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+      context.getResultSender()
+          .lastResult(new CliFunctionResult(member.getId(), false,
+              CliStrings.format(
+                  CliStrings.EXECUTE_FUNCTION__MSG__RESULT_COLLECTOR_0_NOT_FOUND_ERROR_1,
+                  resultCollectorName, e.getMessage())));
+    } catch (Exception e) {
+      logger.error("error executing function " + functionId, e);
       context.getResultSender()
-          .lastResult(new CliFunctionResult(member.getId(), false, ex.getMessage()));
+          .lastResult(new CliFunctionResult(member.getId(), false, "Exception: " + e.getMessage()));
+    } finally {
+      if (loginNeeded) {
+        securityService.logout();
```

Comment:
This could get weird if the `securityService.login` (line 84) fails.  A new `boolean` indicating successful login might be a better gate.

```
81:
boolean loginSuccessful = false;
try{
  if (loginNeeded) { 
    securityService.login(credentials);
    loginSuccessful = true;
  }
...
} finally {
  if (loginSuccessful) {
    securityService.logout();
  }
}
```

Alternatively, if the block in my other comment was a method on the `SecurityService`, we might be able to just check that here: `if (!securityService.loginRequired()){`.   Although I'm not sure I like how that reads, so maybe not.

Reply:
fixed